libs/core/src/main/java/org/elasticsearch/core/internal/provider/EmbeddedImplClassLoader.java

```diff
@@ -136,25 +142,148 @@ public Class<?> findClass(String name) throws ClassNotFoundException {
-        final int size = prefixes.size();
+        Enumeration<URL> enum1 = new Enumeration<>() {
+            private int jarMetaIndex = 0;
+
+            private URL url = null;
+
+            private boolean next() {
+                if (url != null) {
+                    return true;
+                } else {
+                    while (jarMetaIndex < jarMetas.size()) {
+                        URL u = findResourceForPrefixOrNull(name, jarMetas.get(jarMetaIndex));
+                        jarMetaIndex++;
+                        if (u != null) {
+                            url = u;
+                            return true;
+                        }
+                    }
+                    return false;
+                }
+            }
+
+            @Override
+            public boolean hasMoreElements() {
+                return next();
+            }
+
+            @Override
+            public URL nextElement() {
+                if (next() == false) {
+                    throw new NoSuchElementException();
+                }
+                URL u = url;
+                url = null;
+                return u;
+            }
+        };
+
         @SuppressWarnings("unchecked")
-        Enumeration<URL>[] tmp = (Enumeration<URL>[]) new Enumeration<?>[size + 1];
-        for (int i = 0; i < size; i++) {
-            tmp[i] = parent.getResources(prefixes.get(i) + "/" + name);
-        }
-        tmp[size] = parent.getResources(name);
+        Enumeration<URL>[] tmp = (Enumeration<URL>[]) new Enumeration<?>[2];
+        tmp[0] = enum1;
+        tmp[1] = parent.getResources(name);
         return new CompoundEnumeration<>(tmp);
     }
 
+    // -- infra
+
+    private static Map<JarMeta, CodeSource> getProviderPrefixes(ClassLoader parent, String providerName) {
+        String providerPrefix = IMPL_PREFIX + providerName;
+        URL listingURL = parent.getResource(providerPrefix + JAR_LISTING_FILE);
+        if (listingURL == null) {
+            throw new IllegalStateException("missing %s provider jars list".formatted(providerName));
+        }
+        try (
+            InputStream in = listingURL.openStream();
+            InputStreamReader isr = new InputStreamReader(in, StandardCharsets.UTF_8);
+            BufferedReader reader = new BufferedReader(isr)
+        ) {
+            List<String> jars = reader.lines().toList();
+            Map<JarMeta, CodeSource> map = new HashMap<>();
+            for (String jar : jars) {
+                final String jarPrefix = providerPrefix + "/" + jar;
+                JarMeta jam;
+                if (isMultiRelease(parent, jarPrefix)) {
+                    jam = new JarMeta(jarPrefix, true);
+                } else {
+                    jam = new JarMeta(jarPrefix, false);
+                }
+                map.put(jam, codeSource(listingURL, jar));
+            }
+            return Collections.unmodifiableMap(map);
```

Comment:
nit: Is Map.copyOf better to avoid an extra indirection on every access?

Reply:
Done.